prow/plugins/branch-updater/branch-updater.go

```diff
@@ -0,0 +1,148 @@
+	UpdatePullRequestBranch(owner, repo string, number int, expectedHeadSha *string) error
+	GetPullRequest(owner, repo string, number int) (*github.PullRequest, error)
+	GetCombinedStatus(org, repo, ref string) (*github.CombinedStatus, error)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	configInfo := map[string]string{}
+	for _, repo := range enabledRepos {
+		repoUpdateTmpl := "PR branches in repo %s will not be automatically updated."
+		if config.BranchUpdater.ShouldUpdateBranchesForRepo(repo.String()) {
+			repoUpdateTmpl = "PR branches in repo %s will be automatically updated."
+		}
+		configInfo[repo.String()] = fmt.Sprintf(repoUpdateTmpl, repo)
+	}
+
+	yamlSnippet, _ := plugins.CommentMap.GenYaml(&plugins.Configuration{
+		BranchUpdater: plugins.BranchUpdater{
+			DefaultEnabled: true,
+			IgnoredRepos:   []string{"db-improbable/do-not-update-me"},
+			IncludeRepos:   []string{"db-improbable/please-update-me"},
+		},
+	})
+
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The branch-updater plugin updates the branch when an otherwise-mergeable PR is blocked from merging because it's out of date from the HEAD branch",
+		Config:      configInfo,
+		Snippet:     yamlSnippet,
+	}
+	return pluginHelp, nil
+}
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, pullRequestHandler, helpProvider)
+}
+
+func pullRequestHandler(pc plugins.Agent, event github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.Logger, pc.PluginConfig.BranchUpdater, &event)
+}
+
+// If Tide thinks this PR should be mergeable, but GitHub says that the PR's branch is "behind",
+// ask GitHub to update the branch for us.
+func handlePR(gc githubClient, log *logrus.Entry, config plugins.BranchUpdater, event *github.PullRequestEvent) error {
+	// We only care about certain events, so ignore others - this significantly limits the number of race conditions
+	// that can cause multiple actions
+	relevantEvent := false
+	for _, candidate := range enabledEvents {
+		if event.Action == candidate {
+			relevantEvent = true
+		}
+	}
+
+	if !relevantEvent {
+		return nil
+	}
+
+	org := event.Repo.Owner.Login
+	repo := event.Repo.Name
+	pr := event.PullRequest
+
+	// Find out if Tide thinks this PR should be mergeable.
+	statuses, err := gc.GetCombinedStatus(org, repo, pr.Head.Ref)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to get the context statuses on %s/%s#%d.", org, repo, pr.Number)
+		return err
+	}
+
+	// Find the Tide context and bail out if it's not success
+	foundTideContext := false
+	for _, status := range statuses.Statuses {
+		if status.Context == gitHubContextNameTide {
+			if status.State == tideContextStatusSuccess {
+				foundTideContext = true
+				break
+			}
+			return nil
+		}
+	}
+
+	if !foundTideContext {
+		log.Debugf("Skipping PR %d in repo %s/%s not monitored by Tide.", pr.Number, org, repo)
+		return nil
+	}
```

Comment:
similar comment here, extract for loop into own function for readability and shortcutting